File: solutions/java/src/main/java/playground/algorithm/FirstMissingPositive.java
package playground.algorithm;

public class FirstMissingPositive {

    public int firstMissingPositive(int[] nums) {
        int temp;
        for (int i = 0; i < nums.length; ) {
            if (nums[i] - 1 < 0 || nums[i] - 1 >= nums.length || nums[nums[i] - 1] == nums[i]) {
                i++;
                continue;
            }
            temp = nums[nums[i] - 1];
            nums[nums[i] - 1] = nums[i];
            nums[i] = temp;
        }

        for (int j = 0; j < nums.length; j++) {
            if (nums[j] != j + 1) {
                return j + 1;
            }
        }

        return nums.length + 1;
    }
}

File: solutions/go/src/playground/algorithm/first_missing_positive.go
package algorithm

func firstMissingPositive(nums []int) int {
	for i := 0; i < len(nums); {
		if nums[i] < 1 || nums[i] >= len(nums) || nums[i] == nums[nums[i]-1] {
			i++
			continue
		}

		nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
	}

	for i, n := range nums {
		if n != i+1 {
			return i + 1
		}
	}

	return len(nums) + 1
}
